feat: geospatial types support

Oreilles opened the pull request:
This PR implements support for database geometry types. I'll start this message by saying that, although I'd be very proud of making such a contribution to Prisma, I understand that the likeliness of such a massive PR to be accepted is pretty low due to the time I'd take to fully review. How did I end up there ? In the end of last year, I tried to get in touch with the team as I had some experience in the field, and also had the time to work on the feature - but by the time we managed to exchange emails I no longer had enough time to invest in working on this. Last month, I had time and remembered about this project, so I started browing the prisma repo - and realised the ORM core was not actually Typescript, but full Rust, which I had no previous experience with. I questioned my ability to be up to the task, but though: why not have a go at it ? Not having sufficient rust experience, I decided against getting back in touch with the Prisma team and just leverage this project as a learning experience. A month later, although it is very far from being production ready, I believe this branch might deserve some attention which is why I decide to release it. Again, apologies about the size. I worked on this as a learning experience and ended up getting quite ahead of myself.. heh.

Now for the meat of the PR, here's what it includes:

In the `psl` / `query-engine`:

- Two new `ScalarType`s: `Geometry` and `GeoJSON`. Both types refer to the same underlying database types, they only differ in the geometry serialization format the user wants to use: `GeoJSON` for `GeoJSON`, `EWKT` for `Geometry`.  Why do we need both ? `GeoJSON` is the format most people wants to use, as they most likely use TypeScript and `GeoJSON` is the de-facto standard for manipulation geometries on the web. But `Ewkt` allows richer information about the raw geometry: it can be used to serialize geometry in any CRS (when GeoJSON is limited by specification to WGS84), and can also represent more geometry types than geojson (`CIRCULARSTRING  `, `CURVEPOLYGON` etc). As such, only `EWKT` can be safely used to safely & exhaustively read database geometries.
- A new filtering condition: `Geometry(Not)Within` mapping to `st_within`
- A new filtering condition `Geometry(Not)Intersects` mapping to `st_intersects`

In `quaint`: 

- Two new enum `Value`s: `Geometry` and `Geography`, which both holds the same type (`GeometryValue`, a struct with a WKT string and an SRID `i32`  ). We might have needed only one, but MSSQL required making the two explicetly different.
- A new function `geom_from_text`, that maps to the SQL/MM `st_geomfromtext`. 
- A new function `geom_as_text`, that maps to `st_astext.`
- A new visitor function `geom_as_text`, that maps to the `st_astext`
- A new compare function `geometry_is_valid` that maps to `st_isvalid`
- A new compare function `geometry_is_empty` that maps to `st_isempty`
- A new compare function `geometry(_not)_within` that maps to `st_within`
- A new compare function `geometry(_not)_intersects` that maps to `st_intersects`
- A new compare function `geometry_type_equals` that maps to `st_geometrytype(x) = ...`

All in all, what works:

- GeoJSON & EWKT I/O for CockroachDB, MySQL and MSSQL
- GeoJSON & EWKT I/O for PostgreSQL (as long as PostGIS is installed and the public schema is exposed to Prisma)
- GeoJSON & EWKT I/O for SQLite (as long as Spatialite is available and exposed with `SPATIAITE_PATH`)
- GeoJSON I/O for MongoDB
- `Within` and `Intersects` operators for all vendors except MongoDB (Mongo has the `$geoWithin` and `$geoIntersects` query operators, however they cannot be used with the current Prisma implementation, since those are Query operators wherehas Prisma uses Aggregation operators. Unfortunately, they're not compatible, and no geospatial filtering operators are available for aggregate queries in MongoDB.

What hasn't been fully implemented / properly tested:

- Migrations (particularily casts)
- Default value handling
- Index type & operators handling (postgres only ?)

There is still **a lot** of work to do here, but I was quite happy to get the core features (IO and filtering) going and though it was worth sharing. 

My intent with this PR is to show a possible implementation, start a discussion with the team and community about its viability, potential drawbacks and help making these features reach Prisma in a production ready state. Questions about the implementation details are welcome and encouraged. Please get in touch at oreille.nitoref at icloud if you'd like to discuss this more thoroughly!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/prisma-engines?pullRequest=4208) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/prisma-engines?pullRequest=4208) before we can accept your contribution.<br/><hr/>**Aurèle Nitoref** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/prisma-engines?pullRequest=4208) it.</sub>

jordiup commented:
Absolutely boss 🚀

janpio commented:
Amazing.

So that the tests are automatically run going forward:
Can you quickly do a PR to `main` where you change line 5 to `[![Cargo docs](https://github.com/prisma/prisma-engines/actions/workflows/on-push-to-main.yml/badge.svg)](https://github.com/prisma/prisma-engines/actions/workflows/on-push-to-main.yml)`? I can then merge that, and the repository will remember that we trust you and run the tests without me clicking a button. (And that button needs to be fixed anyway 😆)

Oreilles commented:
So there was quite a lot of things I had overlooked when testing on my machine..! I squashed all the fixes and we now have a mostly green test suite 👍

janpio commented:
Two test failures left for now:

1. Something with 1 PostgreSQL metrics test, 7 v.s 9 metrics recorded or something. Could your PR cause this? (We recently change something in metrics, so you might want to rebase (`Update branch` button in GH UI) to get the most recent version of that code).
2. A lot of unhappiness in the Vitess 8 tests with it not being able to introspect the tables because of permission problems. Strange. That looks a bit ... weird, even flaky maybe? I retriggered a test run there. Maybe it just goes green.

PS: No need from our side to combine your PRs and force push, we only merge via "Squash and merge" which combines all the commits, and with the individual commits of the fixes it might be easier to understand what test failures where fixed where - or introduced.

Oreilles commented:
So the two additional queries breaking the tests are due to:

`CREATE EXTENSION IF NOT EXISTS postgis`
https://github.com/Oreilles/prisma-engines/blob/a588e1296288b79df0204397a441f2b3d891b9d9/query-engine/connector-test-kit-rs/qe-setup/src/postgres.rs#L42

and

`SELECT PostGIS_version()`
https://github.com/Oreilles/prisma-engines/blob/a588e1296288b79df0204397a441f2b3d891b9d9/schema-engine/connectors/sql-schema-connector/src/flavour/postgres.rs#L628-L630

Both are needed to run the tests, I updated the counter to 9 to make the test succeed.
( Also rebased on main just in case... )

di-sukharev commented:
How can I help to merge it?

janpio commented:
Test it, figure out how it would fit into Prisma, let the author and use know what else is needed to make this a complete feature.

okonon commented:
@janpio @Oreilles  thanks for all you do guys. I am at the point in my project to start introducing PostGIS and trying to figure out how to test this PR. I cloned @Oreilles feature branch and compiled debug targets of `prisma-engines` how do i tie it into existing `prisma` to run test migration/introspection ? thank you

RafaelKr commented:
@okonon Have a look at the documentation: https://www.prisma.io/docs/concepts/components/prisma-engines#using-custom-engine-libraries-or-binaries

okonon commented:
@RafaelKr thanks a lot for that sorry i missed that section. 
I was able to test introspection from existing database (`prisma db pull`):
<img width="1453" alt="image" src="https://github.com/prisma/prisma-engines/assets/1832398/24204371-a2c7-48c6-8397-e14ec4a370eb">
Looks like it is introspecting correctly.

beardo01 commented:
Adding a comment to voice support for this to be added to Prisma! Looking forward to seeing this merged; keep up the good work.

toniopelo commented:
Definitely would love that to be merged as well :+1:

janpio commented:
Sorry, but this will definitely not be merged as it is only part of the full feature that needs to be implemented.

But we are interested in bringing it there, so let's see what we can do here.

janpio commented:
For now I created a branch based on the commit this was forked from, so that there are no conflicts and @Oreilles can keep committing if they want. 

I also created a local branch of the code that runs all our CI (Buildkite is not run on external contributions), with the goal of publishing the engines to our infrastructure, so we can build a Prisma CLI and Client with them for further testing. 

But unfortunately some tests are failing (results only internally visible): https://github.com/prisma/prisma-engines/pull/4427 
<img width="885" alt="image" src="https://github.com/prisma/prisma-engines/assets/183673/34729a1f-56bd-4959-9e43-7b2cee9aacfe">
Turns out GitHub Actions do not run for these databases, and seems the current code has some problems - and that is blocking my plan to release these engine files :/ 

Random example of a SQLite failure, our of many (`[2023-11-08T21:25:35Z] test result: FAILED. 622 passed; 970 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.73s`):
```
[2023-11-08T21:25:35Z] thread 'writes::top_level_mutations::delete_many_relations::delete_many_rels::pm_c1_no_children' panicked at 'called `Result::unwrap()` on an `Err` value: Error { kind: QueryError(SqlInputError { error: Error { code: Unknown, extended_code: 1 }, msg: "no such function: InitSpatialMetaData", sql: "SELECT InitSpatialMetaData()", offset: 7 }), original_code: Some("1"), original_message: Some("no such function: InitSpatialMetaData") }', /root/build/query-engine/connector-test-kit-rs/qe-setup/src/sqlite.rs:9:65
```

MySQL 5.6 only fails 2:
```
[2023-11-08T21:35:48Z] failures:
[2023-11-08T21:35:48Z]     queries::filters::geometry_filter::geometry_filter_spec::geometric_comparison_filters
[2023-11-08T21:35:48Z]     writes::data_types::native_types::mysql::mysql::native_geojson_srid_geometry_types
```

5.7 only 1 even:
```
[2023-11-08T21:39:12Z] failures:
[2023-11-08T21:39:12Z]     writes::data_types::native_types::mysql::mysql::native_geojson_srid_geometry_types
```

MariaDB the same as 5.6:
```
[2023-11-08T21:31:19Z] failures:
[2023-11-08T21:31:19Z]     queries::filters::geometry_filter::geometry_filter_spec::geometric_comparison_filters
[2023-11-08T21:31:19Z]     writes::data_types::native_types::mysql::mysql::native_geojson_srid_geometry_types
```

Trying to get these tests to run on GitHub Actions now. But maybe you have speculative fixes @Oreilles?

Oreilles commented:
Interesting, as far as I can remember (almost) all tests were succeeding the last time we discussed the PR, and we did have tests for all MySQL versions... though I might remember wrong. Did you resolve merge conflicts when you merge `main`, which could explain tests failing afterwards ?

Anyway, I'll have a look as soon as possible. @janpio if that's something you can do, please send me the test cli commands to reproduce the failing tests.

janpio commented:
Yes, I realize that the `main` merge was not optimal to fully reproduce what was going on here - sorry. I thought about reproducing the state before I did that and make that run in Buildkite, but I couldn't figure out how really.

Oreilles commented:
Should be good now:

- SQLite tests were failing due to an unsuccesfull initialization of Spatialite, which I had unwrapped - it's now oked instead. Tests involving Spatialite will be filtered out later if it is not available.

- The failing write test were failing with MySQL was the one involving geometry columns with srid constraints, which are only available starting with MySQL 8, hence the failing schema initialization.

- The last one is a "fun" one: in MySQL 5.6 and MariaDB, the result of this expression is false:
```sql
select st_intersects(
  st_geomfromtext('POINT(0 0)'),
  st_geomfromtext('POINT(0 0)')
);
```
I replaced the intersection test with one working with all vendors.

janpio commented:
Thanks for even tackling the merge conflicts, that was not expected.

I suggest you add this to your query-engine.yml so that it runs the same tests as Buildkite will: https://github.com/prisma/prisma-engines/pull/4428/files You can see that MySQL (5.7 and 8) and MariaDB fail on GH Actions with 1 test - **but** we should be able to ignore that here and Buildkite would then still pass (I opened an issue to clean this up at https://github.com/prisma/prisma/issues/21896, but no one had time to look at it yet).

Looking at your last commit's results, you can definitely ignore:
- https://github.com/prisma/prisma-engines/actions/runs/6850990006?pr=4208 (we are currently cleaning those up and will have them green soon I hope)
- https://github.com/prisma/prisma-engines/actions/runs/6850990009/job/18626346956?pr=4208 and https://github.com/prisma/prisma-engines/actions/runs/6850990009/job/18626347232?pr=4208 - those are flaky, so I could rerun them a few times to make them pass, but that is ultimately irrelevant

All others are the same singular test that fails: `writes::top_level_mutations::create::geometry_create::create_geometry` - not too bad. I was afraid there was more after your message above.

Oreilles commented:
Turns out it was an much smaller issue that I thought! I squashed the `main` merge to avoid polluting this PR so for reference, this line was the culprit (the null geometry case was seemingly handled elsewhere previously)

https://github.com/Oreilles/prisma-engines/blob/59410f64952de4ed512f8983a4e242b5f327ef5c/quaint/src/visitor.rs#L220

janpio commented:
> I'd be interested in knowing what were the motivation behind the `quaint`'s new `ValueType`

This is what my colleague shared:

> Previously `quaint::Value` was current `quaint::ValueType` it was a enum, which means that if we needed to enrich the representation of a value with metadata for all variants, we needed to modify the source code wherever any variant was referenced. To solve this, and add metadata about a the native column type for postgres (needed to encode information for driver adapters) we promoted `quaint::Value`  from an enum to a struct, and use the old enum as a property of said struct, but renaming the type to  `quiant::ValueType`
> 
> Some more information: https://github.com/prisma/prisma-engines/pull/4311

okonon commented:
thanks for moving this forward.

janpio commented on schema-engine/connectors/sql-schema-connector/src/sql_renderer/sqlite_renderer.rs:
```diff
@@ -141,7 +143,35 @@ impl SqlRenderer for SqliteFlavour {
+                    srid = geom.srid,
+                    ty = geom.ty.as_2d(),
+                    dims = geom.ty.dimensions(),
+                ).unwrap();
```
```suggestion
                )
                .unwrap();
```

janpio commented:
Another look at the failing tests:
- Driver Adapter, LibSQL and SQLite are probably because missing extension? (The LibSQL one we could ignore for now)
- All others are the same test `writes::top_level_mutations::create::geometry_create::create_geometry` caused by `Inconsistent column data: Conversion failed: Value of type Column { catalog: ConstBytes(PhantomData<mysql_common::packets::Catalog>)`
- Vitess ones are flaky and can be ignored

Oreilles commented:
I rebased the branch on main so there shouldn't be any more conflicts for now.
I also managed to add the Spatialite tests to the CI 👍

janpio commented:
Seems like you are getting somewhere.

For the Spatialite problem:
> No space left on device : '/home/ubuntu/actions-runner/_diag/blocks/acc9258d-1293-4903-99d1-cd922e79f8df_4482dee1-f750-5a60-3af2-b3e57514afa7.2'

There are actions that makes more space on the runners: https://github.com/marketplace/actions/maximize-build-disk-space and a few more. But is it expected to take that much space actually?

Only a small number of failing tests on MySQL(like) databases either.

(You can continue ignoring the flakyness in the driver adapter test)

janpio commented:
Thanks for the reminder that our MySQL tests are failing by default, you are of course correct :D

janpio commented:
Here is an integration branch then: https://github.com/prisma/prisma-engines/pull/4507

It shows that the MariaDB failure is indeed the only thing keeping us from an integration version that can be added to `prisma/prisma` to run the tests there.

Is this the relevant failure here?

First one:
```
thread 'queries::filters::field_reference::failure::failure::alphanumeric_json_filter_fails' panicked at query-engine/connectors/sql-query-connector/src/filter/visitor.rs:829:26:
JSON target types only accept strings or numbers, found: {"_ref":"json","_container":"TestModel"}
```
https://github.com/prisma/prisma-engines/actions/runs/7045276806/job/19174664362?pr=4208#step:7:2574
```
thread 'queries::filters::field_reference::failure::failure::alphanumeric_json_filter_fails' panicked at /home/ubuntu/actions-runner/_work/prisma-engines/prisma-engines/query-engine/connector-test-kit-rs/query-tests-setup/src/query_result.rs:106:17:
Expected error with code `P2009` and message `Invalid argument type`, got: `{"errors":[{"error":"JSON target types only accept strings or numbers, found: {\"_ref\":\"json\",\"_container\":\"TestModel\"}","user_facing_error":{"is_panic":true,"message":"JSON target types only accept strings or numbers, found: {\"_ref\":\"json\",\"_container\":\"TestModel\"}","backtrace":null}}]}`
```
https://github.com/prisma/prisma-engines/actions/runs/7045276806/job/19174664362?pr=4208#step:7:2745


And second one:
```
{"errors":[{"error":"Error occurred during query execution:\nConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Server(MysqlError { code: 4079, message: \"Illegal parameter data type varchar for operation 'st_within'\", state: \"HY000\" })), transient: false })","user_facing_error":{"is_panic":false,"message":"Error occurred during query execution:\nConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Server(MysqlError { code: 4079, message: \"Illegal parameter data type varchar for operation 'st_within'\", state: \"HY000\" })), transient: false })","backtrace":null}}]}
```
https://github.com/prisma/prisma-engines/actions/runs/7045276806/job/19174664362?pr=4208#step:7:2935

Oreilles commented:
Yup, that is the one.

janpio commented:
From reading the error message I think it would be more probable that MariaDB has tighter validations, and that MySQL just accepts it with the type it gets? Afaik we do not have a lot of branching to differntiate between MySQL and MariaDB in code.

janpio commented:
Maybe we have a mechanism to exclude MariaDB from running this test? Totally fine to move forward.

Oreilles commented:
Alright then, let's skip it for now.

salvinoto commented:
Is this PostGIS support for prisma?

janpio commented:
@salvinoto Maybe, in the future. For now it is an experiment how a building block could look like that @Oreilles is working on.

javacode123 commented:
@Oreilles hi, could please add some example about how to use GeoJSON and EWKT on this branch. Such as model define、http query on query engine ant etc.

javacode123 commented:
> @Oreilles hi, could please add some example about how to use GeoJSON and EWKT on this branch. Such as model define、http query on query engine ant etc.

i create on model:
```
model places {
  id       Int       @id @default(autoincrement())
  name     String?   @db.VarChar(255)
  location Geometry? @db.Point
}

```
and start query engine, do query:
```
{
    "query": "query GetOneplaces($id: Int!) {\n  data: geo_findFirstplaces(where: {id: {in: [$id]}}) {\n    \n    id\n    name\n    location\n  }\n}",
    "variables": {
        "id": 1
    },
    "operationName": "GetOneplaces"
}
```
it return err:
```
{
    "is_panic": false,
    "message": "Error parsing Graphql query. data did not match any variant of untagged enum GraphqlBody",
    "backtrace": null
}
```

acidjazz commented:
is anyone using this PR? what is left to get done here?

acidjazz commented:
@Oreilles just wanted to say i appreciate your work and loving these commits

Oreilles commented:
Got back at it, now up to date with main!

Made many changes to make it easier to digest (and maintain), specifically:
- Drop support for types incompatible with GeoJSON (M / ZM geometries and extended types (CircularString etc))
- Drop support for (E)WKT and make GeoJSON the only io format
- Therefore, only a single new `Geometry` prisma type identifier is needed

I had to disable the schema engine tests CI for Spatialite, since libspatialite >= 5.0.0 is required and that only ships starting with Ubuntu 22.04. (the CI fail with an SSL error if we try to update the runner).

myishay commented:
I will be so happy when this feature will be merged! Thank you for your work and effort ❤️

ofekBerko commented:
This feature could be a real life changer for me! Thanks for all the work💚💚💚

baba43 commented:
Please push this forward! 👍

acidjazz commented:
> Is development still ongoing for this?

Yes, it is schedule to be completed when George R. R. Martin finishes the book

JCluzet commented:
Please push this forward!

spacerrsdreams commented:
Any updates on this?

mvarchdev commented:
Updates?
